src/aten/sycl/BatchNormKernels.cpp

```diff
@@ -0,0 +1,3703 @@
+#include <ATen/ATen.h>
+#include <ATen/AccumulateType.h>
+#include <ATen/Dispatch.h>
+#include <ATen/native/CanUse32BitIndexMath.h>
+#include <ATen/native/Resize.h>
+#include <ATen/native/TensorIterator.h>
+#include <ATen/xpu/XPUContext.h>
+#include <aten/Resize.h>
+#include <aten/sycl/Loops.h>
+#include <aten/sycl/Reduce.h>
+#include <comm/SYCLContext.h>
+#include <comm/XPUMathCompat.h>
+
+namespace at {
+namespace native {
+namespace xpu {
+
+#define SIMD32 32
+#define SIMD16 16
+
+inline bool batch_norm_use_channels_last_kernels(const at::Tensor& self) {
+  return (
+      self.is_contiguous(at::MemoryFormat::ChannelsLast) ||
+      self.is_contiguous(at::MemoryFormat::ChannelsLast3d) ||
+      (self.is_contiguous() && self.strides()[1] == 1));
+}
+
+struct InvStd {
+  template <typename T>
+  inline T operator()(T var, double epsilon) const {
+    T invstd = 0.0f;
+    if (var != static_cast<T>(0.0f) || epsilon != static_cast<T>(0.0f)) {
+      invstd = static_cast<T>(1.0f) / std::sqrt(var + static_cast<T>(epsilon));
+    }
+    return invstd;
+  }
+};
+
+struct Var {
+  template <typename T>
+  inline T operator()(T var, double epsilon) const {
+    return var;
+  }
+};
+
+static int get_num_threads(int nElem, int max_size) {
+  int threadSizes[6] = {16, 32, 64, 128, 256, max_size};
+  for (int i = 0; i < 6; ++i) {
+    if (nElem <= threadSizes[i]) {
+      return threadSizes[i];
+    }
+  }
+  return max_size;
+}
+
+int get_prefer_wg_size(unsigned int nHw, int simd) {
+  if (nHw < simd)
+    return simd;
+  auto size_problem = get_num_threads(nHw, simd * simd);
+  auto wg_size = syclMaxWorkGroupSize();
+  return std::min(int64_t(size_problem), wg_size);
+}
+
+int get_prefer_simd(int numPlane, int nHw) {
+  // decide SIMD: SIMD32 or SIMD16
+
+  auto dev_id = at::xpu::getDeviceIndexOfCurrentQueue();
+
+  auto* dev_prop = at::xpu::getDeviceProperties(dev_id);
+  auto sub_group_size = dev_prop->sub_group_sizes;
+  int simd = sub_group_size[1];
+  if (simd <= SIMD16)
+    return simd;
+
+  // if max supported simd >16
+  if (nHw <= SIMD16)
+    return SIMD16;
```

Comment:
Here would be a cross-platform compatibility issue.
Conditions,
1. if simd > SIMD16,
2. if all of auto sub_group_size = dev_prop->sub_group_sizes; > SIMD16,
3. nHw <= SIMD16,
Then return SIMD16 is valid.

Reply:
Any updates? Return value should be larger than or equal to `sub_group_size[2]` ?

Reply:
We only tend to support recent common platforms,  not to give golden general codes. Resolved.